plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,194 @@
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}.", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Map<String, QueryGroup> existingQueryGroups = currentClusterState.metadata().queryGroups();
+        String groupName = queryGroup.getName();
+
+        // check if maxQueryGroupCount will breach
+        if (existingQueryGroups.size() + 1 > maxQueryGroupCount) {
+            logger.warn("{} value exceeded its assigned limit of {}.", QUERY_GROUP_COUNT_SETTING_NAME, maxQueryGroupCount);
+            throw new IllegalStateException("Can't create more than " + maxQueryGroupCount + " QueryGroups in the system.");
+        }
+
+        // check for duplicate name
+        Optional<QueryGroup> findExistingGroup = existingQueryGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(groupName))
+            .findFirst();
+        if (findExistingGroup.isPresent()) {
+            logger.warn("QueryGroup with name {} already exists. Not creating a new one.", groupName);
+            throw new IllegalArgumentException("QueryGroup with name " + groupName + " already exists. Not creating a new one.");
+        }
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        Map<ResourceType, Double> totalUsageMap = calculateTotalUsage(existingQueryGroups, queryGroup);
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            if (totalUsageMap.get(resourceType) > 1) {
+                logger.warn("Total resource allocation for {} will go above the max limit of 1.0.", resourceType.getName());
+                throw new IllegalArgumentException(
+                    "Total resource allocation for " + resourceType.getName() + " will go above the max limit of 1.0."
```

Comment:
Are we not supporting percentages? @kaushalmahi12 - can we close on this thread - https://github.com/opensearch-project/OpenSearch/pull/13669/files#r1646712161 ?

Reply:
We can support percentage. Search Backpressure also supports both. In that case we'll need to change the schema from double to string

Reply:
Yes we can.

Reply:
Maybe, we can take percentage as input but internally resolve as ratio? Not sure how Search Backpressure takes care of it

Reply:
SearchBackpressure only takes single value format if it is percentage then it is percentage. But if we store them as ratios then it would not be possible to show them as percentages which in my opinion keeping the single value format is more intuitive from Cx point of view